### PR TITLE
Transient entities in JSON

### LINKED ISSUE #1928: Inconsistencies for transient entities handling in JSON format
Both 4.0 and 4.01 seems to have inconsistencies regarding handling of transient entities in the JSON format.

Null ID for transient entities in JSON:

- In 4.0, it is stated that if an entity is transient, then id must be null in JSON (http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata03/os/odata-json-format-v4.0-errata03-os-complete.html#_Toc453766634).
- In 4.01, it is stated that you don’t even need to return an id in the response (https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#_Toc38457743).

However, it seems to be incompatible with at least 2 other use cases:
1. In the case of recursive navigations (_$expand=Nav($levels=max)_), it is stated you must replace the entity instance by a reference, using its id: 
- http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part1-protocol/odata-v4.0-errata03-os-part1-protocol-complete.html#_Toc453752287
- http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata03/os/odata-json-format-v4.0-errata03-os-complete.html#_Toc453766656
- When the entities are transient, it is impossible since none of them will have an id.
2. In the case of navigationLink, it is stated it must contain the read URL appended with the navigation property name:
- http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata03/os/odata-json-format-v4.0-errata03-os-complete.html#_Toc453766637
- When the entities are transient, they do not have a read URL.

In the case of the XML format, this is not an issue since even transient entities have an id. The client can use those transient ids when parsing and make sense of the references.

### PR DESCRIPTION
**Pull request by ralfhandl**
- [x] Fixes #1928
- [x] Align with Protocol (transient entities may have a read URL)
- [x] Adjusted Protocol (transient entities may have an entity id)

### CODE REVIEW

File: odata-json-format/4 Common Characteristics.md [hunk 1]
```diff
@@ -423,17 +423,18 @@ and have the `null` value. In 4.01 payloads transient
...
-The `id` control information MUST NOT appear for a
-collection. Its meaning in this context is reserved for future versions
-of this specification.
+Services MAY use a string value following the pattern `odata:transient:{some-generated-identifier-unique-within-the-response}` instead of `null`, for example if they  need to solve circular dependencies between transient entities by injecting an entity reference.
 
 Entities with `id` equal to `null` cannot be
 compared to other entities, reread, or updated. If
 [`metadata=minimal`](#metadataminimalodatametadataminimal)
 is specified and the `id` is not present in the entity, then
 the canonical URL MUST be used as the entity-id.
```
Comment (ralfhandl) [inline odata-json-format/4 Common Characteristics.md:432]: This seems to contradict the preceding statement that 

> clients MUST treat entities with neither `id` control
> information nor a full set of key properties as transient entities.

Also it is not clear to me who must do this: the client or the service?

File: odata-json-format/4 Common Characteristics.md [hunk 2]
```diff
@@ -423,17 +423,18 @@ and have the `null` value. In 4.01 payloads transient
...
-of this specification.
+Services MAY use a string value following the pattern `odata:transient:{some-generated-identifier-unique-within-the-response}` instead of `null`, for example if they  need to solve circular dependencies between transient entities by injecting an entity reference.
 
 Entities with `id` equal to `null` cannot be
-compared to other entities, reread, or updated. If
-[`metadata=minimal`](#metadataminimalodatametadataminimal)
+updated or compared to other entities.
+If [`metadata=minimal`](#metadataminimalodatametadataminimal)
 is specified and the `id` is not present in the entity, then
 the canonical URL MUST be used as the entity-id.
```
Comment (ralfhandl) [inline odata-json-format/4 Common Characteristics.md:432]: This seems to contradict the preceding statement that 

> clients MUST treat entities with neither `id` control
> information nor a full set of key properties as transient entities.

Also it is not clear to me who must do this: the client or the service?

**Review by falbert-ptc — APPROVED** [review]

File: odata-json-format/4 Common Characteristics.md [hunk 3]
```diff
@@ -417,23 +417,27 @@ if it does not match convention for the localized key values. If the
...
-
-The `id` control information MUST NOT appear for a
-collection. Its meaning in this context is reserved for future versions
-of this specification.
+Services MAY use a string value following the pattern
+`odata:transient:{some-generated-identifier-unique-within-the-response}`
+instead of `null`, for example if they need to solve circular dependencies
+between transient entities by injecting an entity reference.
 
 Entities with `id` equal to `null` cannot be
```
Comment (ralfhandl) [inline odata-json-format/4 Common Characteristics.md:431]:
```suggestion
Transient entities cannot be
```

File: odata-json-format/4 Common Characteristics.md [hunk 4]
```diff
@@ -417,23 +417,27 @@ if it does not match convention for the localized key values. If the
...
 clients MUST treat entities with neither `id` control
 information nor a full set of key properties as transient entities.
-
-The `id` control information MUST NOT appear for a
-collection. Its meaning in this context is reserved for future versions
-of this specification.
+Services MAY use a string value following the pattern
+`odata:transient:{some-generated-identifier-unique-within-the-response}`
+instead of `null`, for example if they need to solve circular dependencies
+between transient entities by injecting an entity reference.
```
Comment (ralfhandl) [inline odata-json-format/4 Common Characteristics.md:429]: Integrate into previous sentence, make version dependency explicit.

File: odata-json-format/4 Common Characteristics.md [hunk 5]
```diff
@@ -417,23 +417,22 @@ if it does not match convention for the localized key values. If the
 `id` is represented, it MAY be a [relative
 URL](#RelativeURLs).
 
-If the entity is transient (i.e. cannot be read or updated), the
+If the entity is transient (see [OData-Protocol](#ODataProtocol)), the
 `id` control information MUST appear in OData 4.0 payloads
 and have the `null` value. In 4.01 payloads transient
 entities need not have the `id` control information, and 4.01
```
Comment (HeikoTheissen) [inline odata-json-format/4 Common Characteristics.md:423]:
```suggestion
and have the `null` value. In 4.01 or higher payloads transient
entities need not have the `id` control information, and 4.01 or higher
```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: odata-json-format/4 Common Characteristics.md [hunk 6]
```diff
@@ -417,23 +417,22 @@ if it does not match convention for the localized key values. If the
...
 URL](#RelativeURLs).
 
-If the entity is transient (i.e. cannot be read or updated), the
+If the entity is transient (see [OData-Protocol](#ODataProtocol)), the
 `id` control information MUST appear in OData 4.0 payloads
-and have the `null` value. In 4.01 payloads transient
-entities need not have the `id` control information, and 4.01
+and have the `null` value. In 4.01 or greater payloads transient
+entities need not have the `id` control information, and
 clients MUST treat entities with neither `id` control
```
Comment (HeikoTheissen) [inline odata-json-format/4 Common Characteristics.md:424]:
```suggestion
clients receiving such payloads MUST treat entities with neither `id` control
```
If a 4.0 client receives `"@id": null` and a partial key only, is it bound by this statement?

**Review by HeikoTheissen — APPROVED** [review]

**Review by HeikoTheissen — DISMISSED** [review]

File: odata-json-format/1 Introduction.md [hunk 7]
```diff
@@ -25,14 +25,17 @@ Section | Feature / Change | Issue
 Fragment portion of Context URL is not percent-encoded| 
 [368](https://github.com/oasis-tcs/odata-specs/issues/368)
 [Section ##ControlInformationidodataid]| 
-`id` of transient entity can be non-null| 
+`Core.ContentId` in transient entities|
```
Comment (HeikoTheissen) [inline odata-json-format/1 Introduction.md:28]:
```suggestion
`Core.ContentID` in transient entities| 
```

File: odata-json-format/1 Introduction.md [hunk 8]
```diff
@@ -25,14 +25,17 @@ Section | Feature / Change | Issue
...
+`Core.ContentId` in transient entities| 
 [1928](https://github.com/oasis-tcs/odata-specs/issues/1928)
 [Section ##ControlInformationmediaodatamedia]| 
  `mediaContentType` can be `null`| 
 [536](https://github.com/oasis-tcs/odata-specs/issues/536)
 [Section ##StructuralProperty], [Section ##InformativeReferences]| 
 Removed reference to obsolete version of GeoJSON| 
 [456](https://github.com/oasis-tcs/odata-specs/issues/456)
+[Section ##EntityReference]| 
+Reference transient entities via `Core.ContentId`|
```
Comment (HeikoTheissen) [inline odata-json-format/1 Introduction.md:37]:
```suggestion
Reference transient entities via `Core.ContentID`| 
```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: odata-json-format/1 Introduction.md [hunk 9]
```diff
@@ -24,6 +24,9 @@ Section | Feature / Change | Issue
 [Section ##ControlInformationcontextodatacontext]| 
 Fragment portion of Context URL is not percent-encoded| 
 [368](https://github.com/oasis-tcs/odata-specs/issues/368)
+[Section ##ControlInformationidodataid]| 
+Transient entities can be identifiable|
```
Comment (HeikoTheissen) [inline odata-json-format/1 Introduction.md:28]:
```suggestion
Allow `@id` for self-referencing transient entities| 
```

**Review by HeikoTheissen — COMMENTED** [review]
Wordsmithing

File: odata-json-format/4 Common Characteristics.md [hunk 10]
```diff
@@ -419,21 +419,20 @@ URL](#RelativeURLs).
...
-and have the `null` value. In 4.01 payloads transient
-entities need not have the `id` control information, and 4.01
-clients MUST treat entities with neither `id` control
+and have the `null` value. In 4.01 or greater payloads transient
+entities need not have the `id` control information, and
+clients receiving such payloads MUST treat entities with neither `id` control
 information nor a full set of key properties as transient entities.
+In 4.02 payloads transient entities MAY have the `id` control information with a non-null URI value,
+for example to allow solving a circular dependency by injecting an
+[entity reference](#EntityReference) instead of repeating the transient entity.
```
[the remark this answers is not in the document]
Comment (ralfhandl) [reply odata-json-format/4 Common Characteristics.md:428]: We haven't defined or used "self-reference" anywhere.

The current clumsier text uses existing wording from Protocol.

File: odata-protocol/1 Introduction.md [hunk 11]
```diff
@@ -301,8 +301,9 @@ one or both of them may differ from convention.
 
 Transient entities are instances of an entity type that are
 dynamically generated on request and only exist within a response payload.
-They do not possess an entity-id or an update URL and consequently cannot be updated.
-A transient entity may have a read URL, which generates a new transient entity using the same algorithm.
+They do not possess an update URL and consequently cannot be updated.
+A transient entity may have a read URL, which generates a new transient entity using the same algorithm,
+and they may have an entity id if a repeated occurrence in a response needs to be replaced with an entity reference.
```
[the remark this answers is not in the document]
Comment (ralfhandl) [reply odata-protocol/1 Introduction.md:306]: Intentionally used existing (and somewhat clumsier) wording.

**Review by falbert-ptc — APPROVED** [review]

**Pull request merged**